src/api/cpp/nixl_v2_1.h

```diff
@@ -0,0 +1,232 @@
+
+            Space(const struct Local &local);
+            Space(const struct pod::obj::MemSpace &podMemSpace);
+
+            // generate a key for granting access to a local mem space for remote agents
+            uint32_t genPodMemSpaceKeyForLocalMemSpace();
+
+            ~Space();
+        };
+    } // namespace mem
+ 
+    struct MemXferType {
+        mem::Type dst;
+        mem::Type src;
+    };
+
+    namespace container {
+        using BackendType = std::string;
+
+        struct BackendTypeParams {
+            std::vector<MemXferType> supportedMemXferTypes;
+        };
+
+        void getAvailBackendTypes(std::vector<BackendType> &backendTypes);
+
+        void getBackendTypeParams(const struct BackendTypeParams &params);
+    }; // namespace container
+
+    class Agent {
+        enum class PostState {
+            NOT_POSTED,
+            POSTED,
+            COMPLETED
+        };
+
+        enum class PostCompletionStatus {
+            OK,
+            ERROR
+        };
+
+        class MemXfer {
+        public:
+            struct MemXferBuf {
+                class mem::Space space;
+                size_t           ofst;
+                size_t           len;
+            };
+
+            PostState            postState;
+            PostCompletionStatus postCompletionStatus;
+
+            // dst and src must have the same length
+            MemXfer(const std::vector<struct MemXferBuf> &dst,
+                    const std::vector<struct MemXferBuf> &src);
+            
+            MemXfer(const std::vector<struct MemXferBuf> &dst,
+                    const std::vector<struct MemXferBuf> &src,
+                    const std::vector<int>               &dstIndices,
+                    const std::vector<int>               &srcIndices);
+
+            ~MemXfer();
+        };
+
+        class Signal {
+        public:
+            PostState            postSndState;
+            PostCompletionStatus postSndCompletionStatus;
+
+            Signal(const std::string &signalData);
+
+            ~Signal();
+        };
+
+        class pod::Client podClient;
+
+        using RcvSignalCb = std::function<void(class Signal &signal)>;
+
+        Agent(const struct pod::AgentID                 &id,
+              const std::vector<MemXferType>            &memXferTypes,
+              const std::vector<container::BackendType> backendTypes,
+              const struct pod::AgentID                 &podClientServerID,
+              const pod::Client::RcvMsgCb               &podClientRcvMsgCb,
+              const bool                                podClientAutoPoll,
+              const RcvSignalCb                         rcvSignalCb,
+              const bool                                autoPoll)
+              : podClient(podClientServerID, podClientRcvMsgCb, podClientAutoPoll) {}
+
+        ~Agent();
+
+        void postMemXfer(class MemXfer &memXfer);
+
+        void getCompletedMemXfers(std::vector<class MemXfer> &memXfers);
```

Comment:
Passing references to containers is error prone since it is unclear whether the caller is responsible to clear the container before calling the function again or the callee should do that, for example. Return the container instance instead for cleaner, more maintainable interface.

Reply:
We can clarify that the function does NOT clear the container. Returning a container forces a memory allocation.

Reply:
Returning a vector _does not_ force a memory allocation. The resulting vector will be move-constructed by just copying the pointer to data from the source vector instead of deep-copying the contents.

Reply:
The API itself implies quite "heavy" implementation with all the std containers it uses and the amount of features it present to the user. IMO premature micro-optimizations should take lower priority to the simplicity and robustness of the API in a high level C++ application that dynamically allocates memory on a whim (or, more likely, a Python or Rust app communicating with the lib over another layer of bindings on top of it) as opposed to, say, DPDK or kernel API.

Just my 02 cents 🤷‍♂️